File: src/utils/videoId.js
/**
 * 视频ID管理模块
 * 
 * ID 生成规则：基于 dirName:relativePath 生成确定性哈希
 * 同一视频文件始终生成相同的 ID，确保收藏等功能持久有效
 */
const crypto = require('crypto');

// 视频ID映射 (hash ID -> { dirName, relativePath })
const videoIdMap = new Map();

/**
 * 生成确定性视频ID
 * @param {string} dirName 目录名
 * @param {string} relativePath 相对路径
 * @returns {string} 16字符哈希ID
 */
function generateVideoId(dirName, relativePath) {
  const data = `${dirName}:${relativePath}`;
  return crypto.createHash('sha256').update(data).digest('hex').substring(0, 16);
}

/**
 * 获取或创建视频ID
 * @param {string} dirName 目录名
 * @param {string} relativePath 相对路径
 * @returns {string} 视频ID
 */
function getVideoId(dirName, relativePath) {
  // 生成确定性 ID
  const videoId = generateVideoId(dirName, relativePath);
  
  // 存储映射关系（用于反向查找）
  if (!videoIdMap.has(videoId)) {
    videoIdMap.set(videoId, {
      dirName,
      relativePath
    });
  }
  
  return videoId;
}

/**
 * 通过ID获取视频信息
 * @param {string} videoId 视频ID
 * @returns {object|null} 视频信息
 */
function getVideoById(videoId) {
  return videoIdMap.get(videoId) || null;
}

/**
 * 注册视频信息（用于启动时重建映射）
 * @param {string} videoId 视频ID
 * @param {string} dirName 目录名
 * @param {string} relativePath 相对路径
 */
function registerVideo(videoId, dirName, relativePath) {
  videoIdMap.set(videoId, { dirName, relativePath });
}

/**
 * 批量注册视频（扫描完成后调用）
 * @param {Array} videos 视频列表
 */
function registerVideos(videos) {
  videos.forEach(video => {
    const videoId = generateVideoId(video.dirName, video.relativePath);
    videoIdMap.set(videoId, {
      dirName: video.dirName,
      relativePath: video.relativePath
    });
  });
}

module.exports = {
  generateVideoId,
  getVideoId,
  getVideoById,
  registerVideo,
  registerVideos
};

File: go-impl/internal/utils/videoId.go
// Package utils 工具包
package utils

import (
	"crypto/sha256"
	"encoding/hex"
	"sync"

	"video-server/internal/model"
)

// VideoIDManager 视频ID管理器
type VideoIDManager struct {
	idMap sync.Map // ID -> *model.Video
}

var (
	videoIDManager *VideoIDManager
	videoIDOnce    sync.Once
)

// GetVideoIDManager 获取视频ID管理器单例
func GetVideoIDManager() *VideoIDManager {
	videoIDOnce.Do(func() {
		videoIDManager = &VideoIDManager{}
	})
	return videoIDManager
}

// GenerateVideoID 生成确定性视频ID
func GenerateVideoID(dirName, relativePath string) string {
	data := dirName + ":" + relativePath
	hash := sha256.Sum256([]byte(data))
	return hex.EncodeToString(hash[:])[:16]
}

// GetVideoID 获取或创建视频ID
func (vim *VideoIDManager) GetVideoID(dirName, relativePath string) string {
	return GenerateVideoID(dirName, relativePath)
}

// GetVideoByID 通过ID获取视频信息
func (vim *VideoIDManager) GetVideoByID(videoID string) *model.Video {
	if value, ok := vim.idMap.Load(videoID); ok {
		return value.(*model.Video)
	}
	return nil
}

// RegisterVideos 批量注册视频（扫描或加载缓存时调用）
func (vim *VideoIDManager) RegisterVideos(videos []model.Video) {
	for i := range videos {
		videoID := videos[i].ID
		if videoID == "" {
			videoID = GenerateVideoID(videos[i].DirName, videos[i].RelativePath)
			videos[i].ID = videoID
		}
		v := videos[i]
		vim.idMap.Store(videoID, &v)
	}
}
